### PR TITLE
feat: finish batch of engine actions

### PR DESCRIPTION
**Pull request by sileht**
This change to prepare batching for the engine action/command runner.

### CODE REVIEW

File: mergify_engine/context.py [hunk 1]
```diff
@@ -52,6 +52,25 @@
 # https://developer.github.com/v4/enum/mergestatestatus/
 
 
+def get_logger(pull):
```
Comment (jd) [inline mergify_engine/context.py:55]: Why isn't that a method?

File: mergify_engine/tasks/engine/__init__.py [hunk 2]
```diff
@@ -161,24 +161,73 @@ def run(event_type, data):
...
     with github.get_client(owner, repo, installation) as client:
-        _run(client, event_type, data)
+        pull = get_github_pull_from_event(client, event_type, data)
 
+        if not pull:  # pragma: no cover
+            LOG.info(
+                "No pull request found in the event %s, ignoring",
+                event_type,
+                gh_owner=client.owner,
+                gh_repo=client.repo,
```
Comment (jd) [inline mergify_engine/tasks/engine/__init__.py:171]:
```suggestion
                gh_owner=owner,
                gh_repo=repo,
```

or is there any reason to access the client values?

File: mergify_engine/tasks/engine/__init__.py [hunk 3]
```diff
@@ -161,24 +161,73 @@ def run(event_type, data):
...
+        context.get_logger(pull).debug(
+            "No need to proceed queue (got status of an old commit)", event_type,
         )
-        return
+        return True
+    elif (
+        event_type in ["status", "check_suite", "check_run"] and pull["merged"]
+    ):  # pragma: no cover
+        context.get_logger(pull).debug(
+            "No need to proceed queue (got status of a merged pull request)"
```
Comment (jd) [inline mergify_engine/tasks/engine/__init__.py:217]: That does not mean there is not a rule for this though. It sounds like a little premature optimization, no?
(I realize this is already in the code though so we can discuss that later)

File: mergify_engine/tasks/engine/__init__.py [hunk 4]
```diff
@@ -161,24 +161,73 @@ def run(event_type, data):
...
+        for source in sources:
+            if source["event_type"] == "issue_comment":
+                issue_comment_sources.append(source)
+            elif not is_source_obsolete(ctxt.pull, source):
+                ctxt.sources.append(source)
+
+        commands_runner.spawn_pending_commands_tasks(ctxt)
+
+        if issue_comment_sources:
+            for source in issue_comment_sources:
```
Comment (jd) [inline mergify_engine/tasks/engine/__init__.py:191]: Are we sure that the sources are ordered here?

File: mergify_engine/tasks/engine/actions_runner.py [hunk 5]
```diff
@@ -31,13 +31,17 @@
...
-        ):
-            return ""
+    if not (
+        ctxt.pull["merged"]
+        and any(
+            filter(
+                lambda s: s["event_type"] == "pull_request"
+                and s["data"]["action"] == "closed",
+                ctxt.sources,
+            )
```
Comment (jd) [inline mergify_engine/tasks/engine/actions_runner.py:41]:
```suggestion
            s for s in ctxt.sources if s["event_type"] == "pull_request" and s["data"]["action'] == "closed"
```
Not a big deal but it's usually easier to read.

File: mergify_engine/tests/functional/base.py [hunk 6]
```diff
@@ -198,11 +198,15 @@ def _match(cls, data, expected_data):
...
+        r = self._session.get(
             f"https://gh.mergify.io/events-testing?counter={self._counter}",
             data=FAKE_DATA,
             headers={"X-Hub-Signature": "sha1=" + FAKE_HMAC},
-        ).json()
+        )
+        try:
+            return r.json()
+        except json.decoder.JSONDecodeError:
+            raise Exception(r.text)
```
Comment (jd) [inline mergify_engine/tests/functional/base.py:209]:
```suggestion
            raise RuntimeError("Unable to decode JSON", r.text)
```

Comment (sileht) [reply mergify_engine/tests/functional/base.py:209]: That was for debugging, the aioredis bug, I removed it.

File: mergify_engine/tasks/engine/actions_runner.py (hunk 5 above)
Comment (sileht) [reply mergify_engine/tasks/engine/actions_runner.py:41]: done

File: mergify_engine/tasks/engine/__init__.py (hunk 4 above)
Comment (sileht) [reply mergify_engine/tasks/engine/__init__.py:191]: Because of gevent, nothing is ordered already. @sileht

File: mergify_engine/context.py (hunk 1 above)
Comment (sileht) [reply mergify_engine/context.py:55]: It's also used without the context now, I moved it in utils.py

File: mergify_engine/tasks/engine/__init__.py (hunk 2 above)
Comment (sileht) [reply mergify_engine/tasks/engine/__init__.py:171]: No that's the same.

File: mergify_engine/tasks/engine/__init__.py (hunk 3 above)
Comment (sileht) [reply mergify_engine/tasks/engine/__init__.py:217]: I hesitated to removed all of that, it's represent only 1500 events per days. And about this one in particular, maybe people quand status/check rule after pull have been merged.

Comment (sileht) [reply mergify_engine/tasks/engine/__init__.py:217]: I decided to remove them.

File: mergify_engine/tasks/engine/__init__.py (hunk 4 above)
Comment (jd) [reply mergify_engine/tasks/engine/__init__.py:191]: Alright, it might not be a big deal for now but they should be at some point.

**Pull request merged**